cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/cache.go

```diff
@@ -0,0 +1,405 @@
+
+	if !c.watcherStarted {
+		ch, err := c.newResourceWatcherChan()
+		if err != nil {
+			c.watcherMutex.Unlock()
+			log.Errorf("failed to start resource watcher due to: %v", err)
+			return
+		}
+		c.watcherStarted = true
+		c.watcherMutex.Unlock()
+		log.Infof("watcher for [%s] successfully started. waiting for events...", c.config.gvr)
+
+		c.processEvents(ch)
+	} else {
+		c.watcherMutex.Unlock()
+		log.Infof("watcher already started. exiting...")
+	}
+	// we should never reach here under normal usage
+	log.Warningf("+ResourceWatcherCache startResourceWatcher")
+}
+
+func (c *ResourceWatcherCache) newResourceWatcherChan() (<-chan watch.Event, error) {
+	// TODO this is a temp hack to get around the fact that the only clientGetter we've got today
+	// always expects authorization Bearer token in the context
+	ctx := metadata.NewIncomingContext(context.TODO(), metadata.MD{
+		"authorization": []string{"Bearer kaka"},
+	})
+
+	_, dynamicClient, err := c.config.clientGetter(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.FailedPrecondition, "unable to get client due to: %v", err)
+	}
+
+	// this will start a watcher on all namespaces
+	watcher, err := dynamicClient.Resource(c.config.gvr).Namespace("").Watch(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	return watcher.ResultChan(), nil
+}
+
+// this is an infinite loop that waits for new events and processes them when they happen
+func (c *ResourceWatcherCache) processEvents(ch <-chan watch.Event) {
+	for {
+		event := <-ch
+		if event.Type == "" {
+			// not quite sure why this happens (the docs don't say), but it seems to happen quite often
+			continue
+		}
+		log.Infof("got event: type: [%v] object:\n[%s]", event.Type, prettyPrintObject(event.Object))
+		switch event.Type {
+		case watch.Added, watch.Modified, watch.Deleted:
+			unstructuredRepo, ok := event.Object.(*unstructured.Unstructured)
+			if !ok {
+				log.Errorf("Could not cast to unstructured.Unstructured")
+			} else {
+				if event.Type == watch.Added {
+					go c.onAddOrModify(true, unstructuredRepo.Object)
+				} else if event.Type == watch.Modified {
+					go c.onAddOrModify(false, unstructuredRepo.Object)
+				} else {
+					go c.onDelete(unstructuredRepo.Object)
+				}
+			}
+		default:
+			// TODO (gfichtenholt) handle other kinds of events?
+			log.Errorf("got unexpected event: %v", event)
+		}
+	}
+}
+
+// this is effectively a cache PUT operation
+func (c *ResourceWatcherCache) onAddOrModify(add bool, unstructuredObj map[string]interface{}) {
+	defer func() {
+		if c.eventProcessingWaitGroup != nil {
+			c.eventProcessingWaitGroup.Done()
+		}
+	}()
+
+	key, err := c.redisKeyFor(unstructuredObj)
+	if err != nil {
+		log.Errorf("Failed to get redis key due to: %v", err)
+		return
+	}
+
+	// clear that key so cache doesn't contain any stale info for this object if not ready or
+	// indexing or marshalling fails for whatever reason
+	c.redisCli.Del(c.redisCli.Context(), *key)
```

Comment:
> Sorry for taking so long to get through this. I'm happy for this to land but have left a bunch of thoughts.
> 
> The biggest one though, which may be a bit of a blocker - but better to deal with it now than later: I know we'd all talked about redis as an example for the caching layer, and you'd even asked in the past why we use postgresql for caching with the existing helm, but looking at this PR made me realise that cached data in redis may not be easily queried with filters or sorted?
> 
> For example, if this is used to cache the bitnami repo, will it still be a problem to request all pkgs available that begin with "apach" ?

Shouldn't be a problem with the right abstractions, Redis does have rich query capabilities. 
https://redis.io/commands/KEYS
The example you gave should be fairly trivial to implement. Remains to be seen as a whole the query syntax as rich or richer than SQL but I'm sure there's enough there to get us started.

Reply:
> Wouldn't the cache want to have the most recent value if the next update/index/whatever fails? Seems odd to create a time period where there isn't a value for it like this until it does possibly succeed later?

my rationale at the time was to better to have a cache miss than return a out-of-date value? I am willing to be convinced otherwise :-)

Reply:
> Shouldn't be a problem with the right abstractions, Redis does have rich query capabilities.
> https://redis.io/commands/KEYS
> The example you gave should be fairly trivial to implement. Remains to be seen as a whole the query syntax as rich or richer than SQL but I'm sure there's enough there to get us started.

I guess you mean SCAN rather than KEYS (KEYS is for debugging, non-prod use only, afaict)? Checking the guarantees of SCAN, one point in particular (about elements being returned multiple times) could be an issue? See what you think.


> my rationale at the time was to better to have a cache miss than return a out-of-date value? I am willing to be convinced otherwise :-)

Up to you, I was just surprised because elsewhere I saw you explicitly weren't associating expiration with data, so seemed odd to do so here.

Reply:
> > Shouldn't be a problem with the right abstractions, Redis does have rich query capabilities.
> > https://redis.io/commands/KEYS
> > The example you gave should be fairly trivial to implement. Remains to be seen as a whole the query syntax as rich or richer than SQL but I'm sure there's enough there to get us started.
> 
> I guess you mean SCAN rather than KEYS (KEYS is for debugging, non-prod use only, afaict)? Checking the guarantees of SCAN, one point in particular (about elements being returned multiple times) could be an issue? See what you think.
> 
> > my rationale at the time was to better to have a cache miss than return a out-of-date value? I am willing to be convinced otherwise :-)
> 
> Up to you, I was just surprised because elsewhere I saw you explicitly weren't associating expiration with data, so seemed odd to do so here.

just FYI, I am not sure SCAN is the right answer here. I see the redis go client does have 
https://pkg.go.dev/github.com/go-redis/redis/v8@v8.10.0#Client.Keys
which looks like it invokes the KEYS command. It does take a string pattern as an input argument. Nowhere in go client-related documentation (or code) does it tell you it's not meant for production use. I do see a warning about it on https://redis.io/commands/KEYS. I am not sure how relevant that particular warning is for us and our usage. I think it applies to VERY LARGE databases. It does say right above it "While the time complexity for this operation is O(N), the constant times are fairly low. For example, Redis running on an entry level laptop can scan a 1 million key database in 40 milliseconds.". Have we ever encountered a use case with a million keys (i.e. repos in my use case)?

Reply:
> Nowhere in go client-related documentation (or code) does it tell you it's not meant for production use

No, it seems to be just a wrapper around the exact same redis commands. So most functions look undocumented (but their equivalents on the redis site have docs).

But yep, I wasn't worried about the size of our set, more about the functionality itself: it appears to allow filtering on keys (only), not allow filtering/querying based on items within values. You've got more context than me, but it's not clear to me that we'd be able to return, for example, all packages that have an annotation of `Category: foo`. Something like what you can do with json data in postgresql (eg. https://stackoverflow.com/a/42918397 , the [manual](https://www.postgresql.org/docs/13/functions-json.html) has more info, but it's dense without examples).

Reply:
Also, KEYS returns the complete set (without any pagination) while SCAN returns a subset with a cursor (ie. could support pagination) but doesn't guarantee uniqueness (though I don't think this is an issue, in that I'm guessing it doesn't guarantee uniqueness only in the case of insertions and deletions happening while you paginate results). So if we were using KEYS, we'd have to iterate the set to return paginated results anyway. At least, that's the difference I see between KEYS and SCAN.

Reply:
I am going to do one step at a time. First, experiment with KEYS, get a good feel for what the limitations are and then move on to SCAN as necessary. 
That SQL example is kind of unfair, IMHO. That's not even SQL standard, just some advanced postgres-specific extension, albeit useful for us. But I am sure I can find some redis scenarios where postgres falls short

Reply:
Sorry, I'm not trying to make redis look bad or say postgres is better (no interest in that :) ). I'm just keen that we ensure that we can support the current functionality. I'm only pointing out postgresql's json support because that's what we currently use, not because I'm trying to find something better than Redis. For example, see the GenerateWhereClause at https://github.com/kubeapps/kubeapps/blob/master/cmd/assetsvc/pkg/utils/postgresql_utils.go#L211 which uses postgresql's json support for querying the json data which the assetsvc caches.

Reply:
ah, so another way of saying it is that our existing code base isn't portable today. We're locked in to a particular SQL vendor. Bah

Reply:
Well when Kubeapps began from the Monocular project, it was originally a MongoDB document DB. Then when we had to switch away from MongoDB due to licensing issues, we looked for a reliable and trusted OSS solution that would give us the same ability to query JSON blobs. We weren't looking for an SQL solution, but rather something that made an easy transition from MongoDB. Postgresql's JSON support fit that pretty well - that's all.

Reply:
fair enough. And now we need in memory DB with similar JSON support. Redis may or may not fit the bill. I am working on finding out. Just saw this https://oss.redislabs.com/redisjson/#client-libraries - there is an add-on to go client I use. I will find out if it does what we need

Reply:
> > > Shouldn't be a problem with the right abstractions, Redis does have rich query capabilities.
> > > https://redis.io/commands/KEYS
> > > The example you gave should be fairly trivial to implement. Remains to be seen as a whole the query syntax as rich or richer than SQL but I'm sure there's enough there to get us started.
> > 
> > 
> > I guess you mean SCAN rather than KEYS (KEYS is for debugging, non-prod use only, afaict)? Checking the guarantees of SCAN, one point in particular (about elements being returned multiple times) could be an issue? See what you think.
> > > my rationale at the time was to better to have a cache miss than return a out-of-date value? I am willing to be convinced otherwise :-)
> > 
> > 
> > Up to you, I was just surprised because elsewhere I saw you explicitly weren't associating expiration with data, so seemed odd to do so here.
> 
> just FYI, I am not sure SCAN is the right answer here. I see the redis go client does have
> https://pkg.go.dev/github.com/go-redis/redis/v8@v8.10.0#Client.Keys
> which looks like it invokes the KEYS command. It does take a string pattern as an input argument. Nowhere in go client-related documentation (or code) does it tell you it's not meant for production use. I do see a warning about it on https://redis.io/commands/KEYS. I am not sure how relevant that particular warning is for us and our usage. I think it applies to VERY LARGE databases. It does say right above it "While the time complexity for this operation is O(N), the constant times are fairly low. For example, Redis running on an entry level laptop can scan a 1 million key database in 40 milliseconds.". Have we ever encountered a use case with a million keys (i.e. repos in my use case)?

Yikes. I just read https://github.com/redis/redis/issues/3627
I am going to stay away from KEYS :-)